### ISSUE TITLE
Incorrect handling of class that is a subtype of multiple constraints in a value-constrained TypeVar

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#8631]
The following should type check without errors:

```python
from typing import TypeVar, reveal_type

class A:
    ...

class B:
    ...

class C(A, B):
    ...

T = TypeVar("T", A, B)

def test(a: T, b: T):
    ...

test(A(), C())
test(B(), C()) # False positive error
```

**Comment by zingbal** [comment]
reat man min mlbourne n nebi
be partner i hav few to finsh

**Comment by zingbal** [comment]
> The following should type check without errors:
> 
> ```python
> from typing import TypeVar, reveal_type
> 
> class A:
>     ...
> 
> class B:
>     ...
> 
> class C(A, B):
>     ...
> 
> T = TypeVar("T", A, B)
> 
> def test(a: T, b: T):
>     ...
> 
> test(A(), C())
> test(B(), C()) # False positive error
> ```

u doing good bud

**Comment by erictraut** [comment]
I'm going to close this issue for now because the correct behavior of value-constrained TypeVars is poorly specified in the typing spec currently. I'd prefer to first work on getting to an official spec before making any changes in this area and then having to potentially redo them later.

**Comment by lukaspiatkowski** [comment]
Is this one related?

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIAUFQMYA25Azk1AGJhgBcVUfUCZkxoEAsuTgAjYhzBQAvIRIUQACgBEYidNnqANO04HZUAD5QAcmBTEAlDQAmxYFGCdVILoXFSZnW1AAtAB83tp%2B3Lz8rIqyqvb8UCDEMACuIChQrADESgjEUOq6mKwoYPhCSGgo5JL0BTByyWkZsCSFWr6yAAJuYOpAA)

```python
from typing import TypeVar

class Foo:
    pass

TMaybeFoo = TypeVar("TMaybeFoo", Foo, Foo | None)

def foo(r: TMaybeFoo) -> TMaybeFoo:
    s = Foo()
    return s # Type "Foo" is not assignable to return type "TMaybeFoo@foo"
```

**Comment by erictraut** [comment]
It's related only in that there are ambiguities in the spec regarding value-constrained type variables. I'm not optimistic that we'll ever be able to define the behaviors for value-constrained type variables in a way that doesn't conflict with other aspects of the type system. Your code highlights a good example of this. I recommend avoiding value-constrained type variables if you can help it. It's almost always possible to avoid through the use of an overload. In your case, that's a much better solution.
